### ISSUE TITLE
TextInput component lose's ref 

### ISSUE THREAD

**Issue by ashcherbyak** [grommet/grommet#4092]
The use of ref on TextInput does not work as expected. When I try to focus input on first render I get error, because of ref is undefined

```
 const StyledTextInput = styled(TextInput)`
  border: ${({ theme }) => `1px solid ${theme.global.colors.paletteGray10}`};
  padding-top: ${({ theme }) => theme.global.edgeSize.xsmall};
  padding-bottom: ${({ theme }) => theme.global.edgeSize.xsmall};
  border-radius: 0.3rem;
  width: 100%;
`;

const searchRef = useRef();

  useEffect(() => {
    searchRef.current.focus();
  }, []);

    <StyledTextInput
          ref={searchRef}
          type="text"
          onChange={onChangeTextInput}
          value={searchText}
     />
```
When I tryed to use simple ```<input>``` all works as expected 
```
 <input
          ref={searchRef}
          type="text"
          onChange={onChangeTextInput}
          value={searchText}
          placeholder={searchPlaceholder}
          data-automation-id={`${dataAutomationId}-search`}
        />
```
### Expected Behavior
ref should point on TextInput

### Actual Behavior

![image](https://user-images.githubusercontent.com/62097255/82148443-d7040700-985c-11ea-83dc-ca8f49144bc5.png)

### Steps to Reproduce

1. 2. 3.

### Your Environment

- Grommet version: v2.10.0
- Browser Name and version: Chrome
- Operating System and version (desktop or mobile): macOs

**Comment by coltonw** [comment]
This exact issue didn't happen for me in v2.13.0 but now DOES happen in v2.14.0.  It seems like the PR above actually caused this issue to crop up in certain circumstances.

**Comment by ShimiSun** [comment]
@coltonw Can you share a codesandbox that reproduces the issue so we could investigate it further?

**Comment by coltonw** [comment]
https://codesandbox.io/s/cannot-read-property-focus-of-undefined-j0cjp

Seems like it is just failing in our unit tests because our unit tests didn't have a theme wrapper before.  If you wrap it in a Grommet component with a theme, it works fine.  Not sure how defensive you want to code that but figured it was worth noting at least for future developers.

**Comment by IanKBovard** [comment]
I think this issue can be closed. This looks like, to me, an implementation issue. In my opinion, I don't think grommet needs to be defensive about making sure users wrap their app in `<Grommet></Grommet>`, as it is stated pretty clearly on the getting started guide.

**Comment by ShimiSun** [comment]
Thank you @IanKBovard for resolving the issue.

**Comment by lllllllqw** [comment]
@ShimiSun  I found that happened  in class component, even if it already wrapped Grommet component.
It is the codesandbox.
[https://codesandbox.io/s/cannot-read-property-focus-of-undefined-forked-j0mqd?file=/index.js](https://codesandbox.io/s/cannot-read-property-focus-of-undefined-forked-j0mqd?file=/index.js)

**Comment by lllllllqw** [comment]
@ShimiSun  I found that happened  in class component, even if it already wrapped Grommet component.
It is the codesandbox.
[https://codesandbox.io/s/cannot-read-property-focus-of-undefined-forked-j0mqd?file=/index.js](https://codesandbox.io/s/cannot-read-property-focus-of-undefined-forked-j0mqd?file=/index.js)

**Comment by IanKBovard** [comment]
Here is a working example https://codesandbox.io/s/cannot-read-property-focus-of-undefined-forked-bgi81?file=/index.js

**Comment by ShimiSun** [comment]
Thank @IanKBovard! @lllllllqw let us know what you think.

**Comment by lllllllqw** [comment]
> Here is a working example https://codesandbox.io/s/cannot-read-property-focus-of-undefined-forked-bgi81?file=/index.js

Yes, but isn't grommet going to continue to support class-based component?

**Comment by IanKBovard** [comment]
Just posted a pr to continue the conversation. Not sure what the fix should be, but I have a work around for class components.

https://codesandbox.io/s/cannot-read-property-focus-of-undefined-forked-ee063?file=/index.js

**Comment by lllllllqw** [comment]
Thanks! So ref is executed in the class-based component, but the timing may be a bit late.
I feel a bit confused about the behavior of ref.

**Comment by IanKBovard** [comment]
For TextInput the ref is passed through a custom hook when the component mounts. I'm guessing that's why class components never receive the updated ref on mount; which is why using state and ComponentDidUpdate is necessary in this situation.

**Comment by lllllllqw** [comment]
@IanKBovard This approach can indeed solve the ref usage problem in class based components. My question is whether grommet should support class based components perfectly. I think the question has been answered. Thank you very much!

**Comment by ShimiSun** [comment]
Thanks @IanKBovard & @lllllllqw for collaborating on this, @lllllllqw feel free to make code contributions that will help to improve grommet Class component `ref` support.
